File: set1/challenge6.py
# http://cryptopals.com/sets/1/challenges/6/
from functools import partial
from itertools import chain, izip, izip_longest, starmap
from operator import itemgetter, xor

# http://en.wikipedia.org/wiki/Etaoin_shrdlu
english_letter_frequency_lower = {
    'e': 13,
    't': 12,
    'a': 11,
    'o': 10,
    'i': 9,
    'n': 8,
    ' ': 7,
    's': 6,
    'r': 5,
    'h': 4,
    'd': 3,
    'l': 2,
    'u': 1,
}

english_letter_frequency_upper = {l.upper(): f for l, f in english_letter_frequency_lower.iteritems()}

english_letter_frequency = dict(english_letter_frequency_lower)
english_letter_frequency.update(english_letter_frequency_upper)

frequencies = [english_letter_frequency.get(chr(c), 0) for c in range(0, 256)]


def read_file(filename):
    with open(filename, 'r') as f:
        return ''.join([line[:-1] for line in f.readlines()])


def pad(size, bits):
    padding = size - len(bits)
    return padding * [0] + bits


def int2bits(size, integer):
    result = []
    div, mod = divmod(integer, 2)
    result.append(mod)
    while div:
        div, mod = divmod(div, 2)
        result.append(mod)
    return pad(size, list(reversed(result)))


int2bits6 = partial(int2bits, 6)
int2bits8 = partial(int2bits, 8)


def char2intb64(char):
    if 'A' <= char <= 'Z':
        return ord(char) - 65
    elif 'a' <= char <= 'z':
        return ord(char) - 71
    elif '0' <= char <= '9':
        return ord(char) + 4
    elif '+' == char:
        return 62
    elif '/' == char:
        return 63
    else:
        raise ValueError('Invalid base64 char: %s' % char)


def bits2int(bits):
    size = len(bits) - 1
    return sum([b * 2 ** (size - i) for i, b in enumerate(bits)])


def b64(*chars):
    bits = chain(*[int2bits6(char2intb64(c)) for c in chars if c != '='])
    args = [iter(bits)] * 8
    return [bits2int(g) for g in izip(*args) if len(g) == 8]


def b64decode(input):
    args = [iter(input)] * 4
    groups = izip_longest(fillvalue='', *args)
    return chain(*starmap(b64, groups))


def hamming(a, b):
    abits = chain(*map(int2bits8, a))
    bbits = chain(*map(int2bits8, b))

    return sum(starmap(xor, zip(abits, bbits)))


def get_keysize(data):
    candidates = []
    for size in xrange(2, 40):
        args = [iter(data)] * size
        blocks = list(izip_longest(fillvalue=0, *args))
        n = len(blocks) - 1
        distances = sum([hamming(blocks[i], blocks[i+1]) for i in xrange(n)])
        average_distance = distances / float(n)
        normalized_distance = average_distance / size
        candidates.append((size, normalized_distance))

    sorted_candidates = sorted(candidates, key=itemgetter(1))
    return sorted_candidates[0][0]


def xor_list(data, key):
    return [a ^ b for a, b in izip(data, key)]


def score(block, key, size):
    result = sum([frequencies[c] for c in xor_list(block, [key] * size)])
    return (result, key)


def find_key(candidates):
    sorted_candidates = sorted(candidates, key=itemgetter(0), reverse=True)
    best_score, key = sorted_candidates[0]
    return key


def solve_single_xor(block):
    size = len(block)
    candidates = [score(block, key, size) for key in range(0, 256)]
    return find_key(candidates)


def stringify(ints):
    return ''.join([chr(i) for i in ints])


def decrypt(filename):
    # 1. decode the input
    encoded = read_file(filename)
    raw = list(b64decode(encoded))

    # 2. find the most likely key size
    keysize = get_keysize(raw)

    # 3. apply single xor to each transposed block to get the key
    args = [iter(raw)] * keysize
    blocks = izip_longest(fillvalue=0, *args)
    transposed = izip_longest(*blocks)
    key = [solve_single_xor(block) for block in transposed]

    # 4. decrypt the message with this key
    args = [iter(raw)] * keysize
    blocks = list(izip_longest(fillvalue=0, *args))
    decrypted = [stringify(xor_list(block, key)) for block in blocks]
    return ''.join(decrypted)


if __name__ == '__main__':
    input = '6.txt'
    print(decrypt(input))

File: set1/challenge6.hs
-- http://cryptopals.com/sets/1/challenges/6/
import Data.Bits (xor)
import Data.Char (chr, ord, toLower)
import Data.Function (on)
import Data.List (sortBy, transpose)

groupByFill :: Int -> [a] -> a -> [[a]]
groupByFill n [] fill = []
groupByFill n xs fill
    | length xs < n = [xs ++ take (n - length xs) (repeat fill)]
    | otherwise = take n xs:groupByFill n (drop n xs) fill

groupBy :: Int -> [a] -> [[a]]
groupBy n [] = []
groupBy n xs = take n xs:groupBy n (drop n xs)

pad :: Int -> [Int] -> [Int]
pad size xs = if length xs < size then pad size (0:xs) else xs

mods :: Int -> [Int]
mods 0 = []
mods i =
    let (div, mod) = divMod i 2
    in mod:mods div

int2bits :: Int -> Int -> [Int]
int2bits padSize i = pad padSize (reverse (mods i))

int2bits6 = int2bits 6
int2bits8 = int2bits 8

char2intb64 :: Char -> Int
char2intb64 c
  | 'A' <= c && c <= 'Z' = ord(c) - 65
  | 'a' <= c && c <= 'z' = ord(c) - 71
  | '0' <= c && c <= '9' = ord(c) + 4
  | '+' == c = 62
  | '/' == c = 63
  | otherwise   = 0

bits2int :: [Int] -> Int
bits2int xs =
    let size = (length xs) - 1
    in sum [snd(pair) * (2 ^ (size - fst(pair))) | pair <- zip [0..] xs]

b64 :: [Char] -> [Int]
b64 group =
    let bits = concat [int2bits6 (char2intb64 c) | c <- group, c /= '=']
    in [bits2int g | g <- groupBy 8 bits, (length g) == 8]

b64decode :: [Char] -> [Int]
b64decode encoded = concat [b64 g | g <- groupBy 4 encoded]

block2bits :: [Int] -> [Int]
block2bits block = concat [int2bits8 i | i <- block]

hamming :: [Int] -> [Int] -> Int
hamming a b = sum $ zipWith xor (block2bits a) (block2bits b)

distances :: [[Int]] -> [Int]
distances [] = []
distances [x] = []
distances (h:xs) = (hamming h (head xs)):distances xs

normalizedDistance :: Int -> [Int] -> Float
normalizedDistance size input =
    let blocks = groupByFill size input 0
        sumDistances = sum $ distances blocks
        n = length blocks - 1
    in ((fromIntegral sumDistances) / (fromIntegral n)) / (fromIntegral size)

findKeySize :: [Int] -> Int
findKeySize input =
    let candidates = [(size, normalizedDistance size input) | size <- [2..40]]
        sorted = sortBy (compare `on` snd) candidates
    in fst $ head sorted

frequency :: Int -> Int
frequency byte
    | char == 'e' = 13
    | char == 't' = 12
    | char == 'a' = 11
    | char == 'o' = 10
    | char == 'i' = 9
    | char == 'n' = 8
    | char == ' ' = 7
    | char == 's' = 6
    | char == 'r' = 5
    | char == 'h' = 4
    | char == 'd' = 3
    | char == 'l' = 2
    | char == 'u' = 1
    | otherwise   = 0
    where char = toLower $ chr byte

xorList :: [Int] -> [Int] -> [Int]
xorList = zipWith xor

score :: [Int] -> Int -> Int -> (Int, Int)
score block key size = (sum [frequency c | c <- xorList block (repeat key)], key)

solveSingleXOR :: [Int] -> Int
solveSingleXOR block =
    let size = length block
        candidates = [score block key size | key <- [0..255]]
        sortedCandidates = sortBy (flip (compare `on` fst)) candidates
    in snd $ head sortedCandidates

findKey :: [Int] -> Int -> [Int]
findKey input keySize =
    let blocks = transpose $ groupByFill keySize input 0
    in [solveSingleXOR block | block <- blocks]

stringify :: [Int] -> [Char]
stringify list = [chr d | d <- list]

xorDecrypt :: [Int] -> Int -> [Int] -> [Int]
xorDecrypt input keySize key = concat [xorList block key | block <- groupByFill keySize input 0]

decrypt :: [Char] -> [Char]
decrypt b64 =
    let raw  = b64decode [c | c <- b64, c /= '\n']
        keySize = findKeySize raw
        key = findKey raw keySize
    in stringify $ xorDecrypt raw keySize key

main = do
    contents <- readFile "6.txt"
    putStrLn $ decrypt contents
